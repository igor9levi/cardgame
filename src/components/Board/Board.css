.board {
    display: flex;
    flex-wrap: nowrap;
    flex: 1;
    flex-direction: row;
    justify-content: flex-end;
    align-items: center;
    /*width: 100%;*/
    /*height: 100%;*/
    font-size: 100px;
    color: #00f;
}

.player {
    /*flex: 1;*/
    width: 300px;
    height: 200px;
}

.player-cards {
     /*flex: 1;*/
     /*flex-shrink: 0;*/
     max-width: 100%;
     width: 50px;
     height: 50px;

 }

.wrapper {
    display: flex;
    flex: 1;
    flex-direction: column;
    background-color: indianred;
    height: 100%;
    /*max-height: 100%;*/
    /*max-width: 100%;*/
    width: 100%;
}

.player {
    background-color: cadetblue;
}

.middle-row {
    /*display: flex;*/
    /*flex-direction: row;*/
    display: block;
    width: 100%;
}

.middle-row > .player {
    /*display: flex;*/
    /*flex: 1;*/
    /*display: block;*/
    display: inline-block;
    width: 30%;
}

.middle-row > .table {
    /*display: flex;*/
    /*flex: 1;*/
    background-color: aquamarine;
    display: inline-block;
    width: 30%;
}

.top-row {
    height: 20%;
    background-color: blue;
}

.bottom-row {
    height: 20%;
    background-color: blue;
}

.player-top {
    /*flex: 1;*/
    background-color: blueviolet;
    width: 100%;
}

.player-bottom {
    /*flex: 1;*/
    background-color: olivedrab;
    width: 100%;
}

.player-left {
    /*flex: 1;*/
}

.player-right {
    /*flex: 1;*/
}