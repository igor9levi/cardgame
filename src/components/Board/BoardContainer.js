import { connect } from 'react-redux';
import BoardComponent from './Board';
import { getCards } from '../../modules/Player/playerSelectors';
import {addCardToTable, flushTable, setRoundWinner} from '../../modules/Player/playerActions';
import { setEndStatus } from '../App/appActions';

export const mapStateToProps = state => ({
  cards: getCards(state),
  numPlayers: state.app.numPlayers,
  score: state.player.score,
  table: state.player.table,
});

export default connect(mapStateToProps, { addCardToTable, setRoundWinner, setEndStatus, flushTable })(BoardComponent);
