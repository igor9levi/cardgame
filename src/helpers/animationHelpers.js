export const pause = (numSeconds = 2000) => (new Promise((resolve) => {
  setTimeout(() => resolve('ok'), numSeconds);
}));
