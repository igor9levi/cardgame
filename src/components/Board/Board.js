import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import './Board.css';
import Card from './Card';
import {calculateRoundWinner} from '../../helpers/roundHelpers';

class Board extends PureComponent {
  static propTypes = {
    cards: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.object)).isRequired,
    numPlayers: PropTypes.number.isRequired,
    table: PropTypes.arrayOf(PropTypes.object),
    score: PropTypes.arrayOf(PropTypes.number),
    addCardToTable: PropTypes.func.isRequired,
    flushTable: PropTypes.func.isRequired,
  }

  static defaultProps = {
    table: [],
    score: [],
  }

  constructor(props) {
    super(props);
    this.initialState = {
      center: {},
    };
    this.state = this.initialState;

    this.playerTurns = [...Array(props.numPlayers).keys()];

    this.refTable = React.createRef();
    props.cards.map(item => item.map((card) => {
      const ref = React.createRef();
      this[card.code] = ref;
      return ref;
    }));
  }

  componentDidMount() {
    const center = this.getCenter();
    this.setState({
      center,
    });
  }

  shufflePlayers = () => {
    const { playerTurns } = this;
    this.playerTurns = playerTurns.slice(1).concat(playerTurns[0]);
  }

  computerPlay = () => {
    const { cards, addCardToTable } = this.props;
    const { playerTurns } = this;

    // Todo: pick random number between 0 and playerTurns[0].length-1
    const randomNumber = 5;
    const card = cards[playerTurns[0]][randomNumber];
    const { playerId, value, code } = card;
    addCardToTable({ playerId, value, code });
  }

  startNewRound = () => {
    // Todo: calculate round winner, update player score (winnerCard.playerId), dispatch flush table
    const winnerCard = calculateRoundWinner(this.props.table);
    this.props.flushTable();
  }

  playRounds = () => {
    this.shufflePlayers();

    while (this.playerTurns[0] !== 0) {
      this.computerPlay();
      if (this.props.table.length === this.props.numPlayers) {
        this.startNewRound();
        if (this.props.cards.length === 0) {
          // Todo: Dispatch end game
        }
      } else {
        this.shufflePlayers();
      }
    }
  }

  renderPlayer = ({ cards, player }) => {
    if (!cards || !cards[player]) return null;

    const { center } = this.state;

    return cards[player].map(card => (
      <Card
        reference={this[card.code]}
        key={card.code}
        code={card.code}
        alt={card.value}
        src={card.image}
        center={center}
        value={card.value}
        playerId={card.playerId}
        playRounds={this.playRounds}
      />
    ));
  }

  getCenter = () => {
    const node = this.refTable.current;
    const centerX = node.offsetLeft + node.offsetWidth / 2;
    const centerY = node.offsetTop + node.offsetHeight / 2;
    const center = {
      centerX,
      centerY,
    };

    return center;
  }

  render() {
    const { cards } = this.props;

    // Todo: add show each individual score
    return (
      <div className="wrapper">
        <div className="top-row">
          {this.renderPlayer({ player: 2, cards })}
        </div>
        <div className="middle-row">
          <div className="player player-left">
            {this.renderPlayer({ player: 1, cards })}
          </div>
          <div className="player table" ref={this.refTable} />
          <div className="player player-right">
            {this.renderPlayer({ player: 3, cards })}
          </div>
        </div>
        <div className="bottom-row">
          {this.renderPlayer({ player: 0, cards })}
        </div>
      </div>
    );
  }
}

export default Board;
