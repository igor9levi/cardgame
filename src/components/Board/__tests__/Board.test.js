import React from 'react';
import { shallow } from 'enzyme';
import Board from '../Board';
import * as roundHelpers from '../../../helpers/roundHelpers';
import * as animationHelpers from '../../../helpers/animationHelpers';

const testCards = [
  [{
    key: 1,
    code: 1,
    alt: 1,
    src: 'img',
    value: 1,
    playerId: 0,
  }], [{
    key: 2,
    code: 2,
    alt: 2,
    src: 'img',
    value: 2,
    playerId: 1,
  }], [{
    key: 3,
    code: 3,
    alt: 3,
    src: 'img',
    value: 3,
    playerId: 2,
  }],
];

const props = {
  cards: testCards,
  numPlayers: 3,
  table: [{}, {}],
  score: [0, 0, 0],
  addCardToTable: jest.fn(),
  flushTable: jest.fn(),
  setEndStatus: jest.fn(),
  setRoundWinner: jest.fn(),
  unblockClick: jest.fn(),
};

roundHelpers.getCenter = jest.fn(() => ({
  centerX: 0,
  centerY: 0,
}));

describe('Test Board component', () => {
  describe('Test rendering', () => {
    it('Renders correctly', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('Renders correctly all 4 players', () => {
      const wrapper = shallow(
        <Board {...props} numPlayers={4} score={[0, 0, 0, 0]} />,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Test removeCard', () => {
    it('Adds card to purgatory', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      wrapper.instance().purgatory = [2];
      wrapper.instance().removeCard(4);
      expect(wrapper.instance().purgatory).toEqual([2, 4]);
    });

    it('Not to add card to purgatory if already exists', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      wrapper.instance().purgatory = [2, 3];
      wrapper.instance().removeCard(2);
      expect(wrapper.instance().purgatory).toEqual([2, 3]);
    });

    it('Adds card to purgatory and flushes table', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      wrapper.instance().purgatory = [2, 3];
      wrapper.instance().removeCard(4);
      expect(wrapper.instance().purgatory).toEqual([]);
      expect(wrapper.instance().props.flushTable).toBeCalled();
      expect(wrapper.instance().props.flushTable).toHaveBeenCalledTimes(1);
    });
  });

  describe('Test setWinnerToPlay', () => {
    it('shift playersTurn to match winner and call setState', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      wrapper.instance().setState = jest.fn();
      wrapper.instance().state.playersTurn = [0, 1, 2, 3];
      wrapper.instance().setWinnerToPlay(2);
      expect(wrapper.instance().setState).toBeCalled();
      expect(wrapper.instance().setState)
        .toHaveBeenCalledWith({ playersTurn: [2, 3, 0, 1] });
    });
  });

  describe('Test shufflePlayers', () => {
    it('rotate playersTurn by one and call setState', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      wrapper.instance().setState = jest.fn();
      wrapper.instance().state = { playersTurn: [0, 1, 2, 3] };
      wrapper.instance().shufflePlayers();
      expect(wrapper.instance().setState).toBeCalled();
      expect(wrapper.instance().setState)
        .toHaveBeenCalledWith({ playersTurn: [1, 2, 3, 0] });
    });
  });

  describe('Test componentDidUpdate', () => {
    it('test call endGame when game finishes', () => {
      const prevProps = {};
      const prevState = {};
      const wrapper = shallow(
        <Board {...props} />,
      );
      wrapper.instance().playRound = jest.fn();
      roundHelpers.gameEnd = jest.fn(() => false);
      roundHelpers.shouldPlayRound = jest.fn(() => true);
      wrapper.instance().componentDidUpdate(prevProps, prevState);
      expect(roundHelpers.gameEnd).toBeCalled();
      expect(wrapper.instance().props.setEndStatus).not.toBeCalled();
      expect(roundHelpers.shouldPlayRound).toBeCalled();
      expect(wrapper.instance().playRound).toBeCalled();
    });

    it('test call endGame when game finishes', () => {
      const prevProps = {};
      const prevState = {};
      const wrapper = shallow(
        <Board {...props} />,
      );
      roundHelpers.gameEnd = jest.fn(() => true);
      roundHelpers.shouldPlayRound = jest.fn(() => false);
      wrapper.instance().componentDidUpdate(prevProps, prevState);
      expect(roundHelpers.gameEnd).toBeCalled();
      expect(wrapper.instance().props.setEndStatus).toBeCalled();
    });
  });

  describe('Test resetRound', () => {
    it('Calls setWinnerToPlay and setRoundWinner', () => {
      const wrapper = shallow(
        <Board {...props} />,
      );
      roundHelpers.calculateRoundWinner = jest.fn(() => ({ playerId: 1 }));
      wrapper.instance().setWinnerToPlay = jest.fn();
      wrapper.instance().resetRound();
      expect(wrapper.instance().props.setRoundWinner).toBeCalled();
      expect(wrapper.instance().setWinnerToPlay).toBeCalled();
    });
  });

  describe('Test playRound', () => {
    it('resets round', async () => {
      const wrapper = shallow(
        <Board {...props} numPlayers={2} />,
      );
      wrapper.instance().state = { playersTurn: [0, 1] };
      wrapper.instance().resetRound = jest.fn();
      const result = await wrapper.instance().playRound();
      expect(wrapper.instance().resetRound).toBeCalled();
      expect(result).toEqual(undefined);
    });

    it('it is humans turn to play', async () => {
      const wrapper = shallow(
        <Board {...props} numPlayers={3} />,
      );
      animationHelpers.pause = jest.fn(() => Promise.resolve({}));
      // wrapper.instance().state = { playersTurn: [0, 1, 2] };
      wrapper.setState({ playersTurn: [0, 1, 2] });
      wrapper.instance().resetRound = jest.fn();
      const result = await wrapper.instance().playRound();

      expect(result).toEqual(false);
      expect(wrapper.instance().resetRound).not.toBeCalled();
      expect(animationHelpers.pause).not.toBeCalled();
      expect(props.unblockClick).toBeCalled();
    });

    it('add random card to table after 500ms', async () => {
      const wrapper = shallow(
        <Board {...props} numPlayers={3} />,
      );
      animationHelpers.pause = jest.fn(() => Promise.resolve({}));
      wrapper.instance().state = { playersTurn: [1, 2, 0] };
      wrapper.instance().resetRound = jest.fn();
      const result = await wrapper.instance().playRound();
      expect(wrapper.instance().resetRound).not.toBeCalled();
      expect(animationHelpers.pause).toBeCalled();
      expect(wrapper.instance().props.addCardToTable).toBeCalled();
      expect(result).toEqual(undefined);
    });
  });
});
