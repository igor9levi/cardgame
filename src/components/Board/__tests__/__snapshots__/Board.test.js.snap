// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Board component Test rendering Renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Board
    addCardToTable={[Function]}
    cards={
        Array [
            Array [
              Object {
                "alt": 1,
                "code": 1,
                "key": 1,
                "playerId": 0,
                "src": "img",
                "value": 1,
              },
            ],
            Array [
              Object {
                "alt": 2,
                "code": 2,
                "key": 2,
                "playerId": 1,
                "src": "img",
                "value": 2,
              },
            ],
            Array [
              Object {
                "alt": 3,
                "code": 3,
                "key": 3,
                "playerId": 2,
                "src": "img",
                "value": 3,
              },
            ],
          ]
    }
    flushTable={[Function]}
    numPlayers={3}
    score={
        Array [
            0,
            0,
            0,
          ]
    }
    setEndStatus={[Function]}
    setRoundWinner={[Function]}
    table={
        Array [
            Object {},
            Object {},
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="player player-left"
>
          <p>
                    Score:
                    0
          </p>
          <Connect(Card)
                    alt={2}
                    animationFinished={[Function]}
                    blockClick={true}
                    center={
                              Object {
                                        "centerX": 0,
                                        "centerY": 0,
                                      }
                    }
                    code={2}
                    playRounds={undefined}
                    playerId={1}
                    removeCard={[Function]}
                    src={undefined}
                    value={2}
          />
</div>,
        <div
          className="middle-row"
>
          <div>
                    <div
                              className="top-row"
                    >
                              <p>
                                        Score:
                                        0
                              </p>
                              <Connect(Card)
                                        alt={3}
                                        animationFinished={[Function]}
                                        blockClick={true}
                                        center={
                                                  Object {
                                                            "centerX": 0,
                                                            "centerY": 0,
                                                          }
                                        }
                                        code={3}
                                        playRounds={undefined}
                                        playerId={2}
                                        removeCard={[Function]}
                                        src={undefined}
                                        value={3}
                              />
                    </div>
                    <div
                              className="player table"
                    />
                    <div
                              className="bottom-row"
                    >
                              <p>
                                        Score:
                                        0
                              </p>
                              <Connect(Card)
                                        alt={1}
                                        animationFinished={[Function]}
                                        blockClick={false}
                                        center={
                                                  Object {
                                                            "centerX": 0,
                                                            "centerY": 0,
                                                          }
                                        }
                                        code={1}
                                        playRounds={undefined}
                                        playerId={0}
                                        removeCard={[Function]}
                                        src={undefined}
                                        value={1}
                              />
                    </div>
          </div>
</div>,
        <div
          className="player player-right"
>
          
</div>,
      ],
      "className": "wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              Score:
              0
</p>,
            Array [
              <Connect(Card)
                alt={2}
                animationFinished={[Function]}
                blockClick={true}
                center={
                                Object {
                                                "centerX": 0,
                                                "centerY": 0,
                                              }
                }
                code={2}
                playRounds={undefined}
                playerId={1}
                removeCard={[Function]}
                src={undefined}
                value={2}
/>,
            ],
          ],
          "className": "player player-left",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Score:",
                0,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Score:",
              0,
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "alt": 2,
              "animationFinished": [Function],
              "blockClick": true,
              "center": Object {
                "centerX": 0,
                "centerY": 0,
              },
              "code": 2,
              "playRounds": undefined,
              "playerId": 1,
              "removeCard": [Function],
              "src": undefined,
              "value": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <div
                        className="top-row"
            >
                        <p>
                                    Score:
                                    0
                        </p>
                        <Connect(Card)
                                    alt={3}
                                    animationFinished={[Function]}
                                    blockClick={true}
                                    center={
                                                Object {
                                                            "centerX": 0,
                                                            "centerY": 0,
                                                          }
                                    }
                                    code={3}
                                    playRounds={undefined}
                                    playerId={2}
                                    removeCard={[Function]}
                                    src={undefined}
                                    value={3}
                        />
            </div>
            <div
                        className="player table"
            />
            <div
                        className="bottom-row"
            >
                        <p>
                                    Score:
                                    0
                        </p>
                        <Connect(Card)
                                    alt={1}
                                    animationFinished={[Function]}
                                    blockClick={false}
                                    center={
                                                Object {
                                                            "centerX": 0,
                                                            "centerY": 0,
                                                          }
                                    }
                                    code={1}
                                    playRounds={undefined}
                                    playerId={0}
                                    removeCard={[Function]}
                                    src={undefined}
                                    value={1}
                        />
            </div>
</div>,
          "className": "middle-row",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="top-row"
>
                <p>
                                Score:
                                0
                </p>
                <Connect(Card)
                                alt={3}
                                animationFinished={[Function]}
                                blockClick={true}
                                center={
                                                Object {
                                                                "centerX": 0,
                                                                "centerY": 0,
                                                              }
                                }
                                code={3}
                                playRounds={undefined}
                                playerId={2}
                                removeCard={[Function]}
                                src={undefined}
                                value={3}
                />
</div>,
              <div
                className="player table"
/>,
              <div
                className="bottom-row"
>
                <p>
                                Score:
                                0
                </p>
                <Connect(Card)
                                alt={1}
                                animationFinished={[Function]}
                                blockClick={false}
                                center={
                                                Object {
                                                                "centerX": 0,
                                                                "centerY": 0,
                                                              }
                                }
                                code={1}
                                playRounds={undefined}
                                playerId={0}
                                removeCard={[Function]}
                                src={undefined}
                                value={1}
                />
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p>
                    Score:
                    0
</p>,
                  Array [
                    <Connect(Card)
                      alt={3}
                      animationFinished={[Function]}
                      blockClick={true}
                      center={
                                            Object {
                                                                  "centerX": 0,
                                                                  "centerY": 0,
                                                                }
                      }
                      code={3}
                      playRounds={undefined}
                      playerId={2}
                      removeCard={[Function]}
                      src={undefined}
                      value={3}
/>,
                  ],
                ],
                "className": "top-row",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Score:",
                      0,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "Score:",
                    0,
                  ],
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": "3",
                  "nodeType": "class",
                  "props": Object {
                    "alt": 3,
                    "animationFinished": [Function],
                    "blockClick": true,
                    "center": Object {
                      "centerX": 0,
                      "centerY": 0,
                    },
                    "code": 3,
                    "playRounds": undefined,
                    "playerId": 2,
                    "removeCard": [Function],
                    "src": undefined,
                    "value": 3,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "player table",
              },
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p>
                    Score:
                    0
</p>,
                  Array [
                    <Connect(Card)
                      alt={1}
                      animationFinished={[Function]}
                      blockClick={false}
                      center={
                                            Object {
                                                                  "centerX": 0,
                                                                  "centerY": 0,
                                                                }
                      }
                      code={1}
                      playRounds={undefined}
                      playerId={0}
                      removeCard={[Function]}
                      src={undefined}
                      value={1}
/>,
                  ],
                ],
                "className": "bottom-row",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Score:",
                      0,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "Score:",
                    0,
                  ],
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "class",
                  "props": Object {
                    "alt": 1,
                    "animationFinished": [Function],
                    "blockClick": false,
                    "center": Object {
                      "centerX": 0,
                      "centerY": 0,
                    },
                    "code": 1,
                    "playRounds": undefined,
                    "playerId": 0,
                    "removeCard": [Function],
                    "src": undefined,
                    "value": 1,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            null,
          ],
          "className": "player player-right",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          null,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="player player-left"
>
            <p>
                        Score:
                        0
            </p>
            <Connect(Card)
                        alt={2}
                        animationFinished={[Function]}
                        blockClick={true}
                        center={
                                    Object {
                                                "centerX": 0,
                                                "centerY": 0,
                                              }
                        }
                        code={2}
                        playRounds={undefined}
                        playerId={1}
                        removeCard={[Function]}
                        src={undefined}
                        value={2}
            />
</div>,
          <div
            className="middle-row"
>
            <div>
                        <div
                                    className="top-row"
                        >
                                    <p>
                                                Score:
                                                0
                                    </p>
                                    <Connect(Card)
                                                alt={3}
                                                animationFinished={[Function]}
                                                blockClick={true}
                                                center={
                                                            Object {
                                                                        "centerX": 0,
                                                                        "centerY": 0,
                                                                      }
                                                }
                                                code={3}
                                                playRounds={undefined}
                                                playerId={2}
                                                removeCard={[Function]}
                                                src={undefined}
                                                value={3}
                                    />
                        </div>
                        <div
                                    className="player table"
                        />
                        <div
                                    className="bottom-row"
                        >
                                    <p>
                                                Score:
                                                0
                                    </p>
                                    <Connect(Card)
                                                alt={1}
                                                animationFinished={[Function]}
                                                blockClick={false}
                                                center={
                                                            Object {
                                                                        "centerX": 0,
                                                                        "centerY": 0,
                                                                      }
                                                }
                                                code={1}
                                                playRounds={undefined}
                                                playerId={0}
                                                removeCard={[Function]}
                                                src={undefined}
                                                value={1}
                                    />
                        </div>
            </div>
</div>,
          <div
            className="player player-right"
>
            
</div>,
        ],
        "className": "wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                Score:
                0
</p>,
              Array [
                <Connect(Card)
                  alt={2}
                  animationFinished={[Function]}
                  blockClick={true}
                  center={
                                    Object {
                                                      "centerX": 0,
                                                      "centerY": 0,
                                                    }
                  }
                  code={2}
                  playRounds={undefined}
                  playerId={1}
                  removeCard={[Function]}
                  src={undefined}
                  value={2}
/>,
              ],
            ],
            "className": "player player-left",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Score:",
                  0,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Score:",
                0,
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "alt": 2,
                "animationFinished": [Function],
                "blockClick": true,
                "center": Object {
                  "centerX": 0,
                  "centerY": 0,
                },
                "code": 2,
                "playRounds": undefined,
                "playerId": 1,
                "removeCard": [Function],
                "src": undefined,
                "value": 2,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <div
                            className="top-row"
              >
                            <p>
                                          Score:
                                          0
                            </p>
                            <Connect(Card)
                                          alt={3}
                                          animationFinished={[Function]}
                                          blockClick={true}
                                          center={
                                                        Object {
                                                                      "centerX": 0,
                                                                      "centerY": 0,
                                                                    }
                                          }
                                          code={3}
                                          playRounds={undefined}
                                          playerId={2}
                                          removeCard={[Function]}
                                          src={undefined}
                                          value={3}
                            />
              </div>
              <div
                            className="player table"
              />
              <div
                            className="bottom-row"
              >
                            <p>
                                          Score:
                                          0
                            </p>
                            <Connect(Card)
                                          alt={1}
                                          animationFinished={[Function]}
                                          blockClick={false}
                                          center={
                                                        Object {
                                                                      "centerX": 0,
                                                                      "centerY": 0,
                                                                    }
                                          }
                                          code={1}
                                          playRounds={undefined}
                                          playerId={0}
                                          removeCard={[Function]}
                                          src={undefined}
                                          value={1}
                            />
              </div>
</div>,
            "className": "middle-row",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="top-row"
>
                  <p>
                                    Score:
                                    0
                  </p>
                  <Connect(Card)
                                    alt={3}
                                    animationFinished={[Function]}
                                    blockClick={true}
                                    center={
                                                      Object {
                                                                        "centerX": 0,
                                                                        "centerY": 0,
                                                                      }
                                    }
                                    code={3}
                                    playRounds={undefined}
                                    playerId={2}
                                    removeCard={[Function]}
                                    src={undefined}
                                    value={3}
                  />
</div>,
                <div
                  className="player table"
/>,
                <div
                  className="bottom-row"
>
                  <p>
                                    Score:
                                    0
                  </p>
                  <Connect(Card)
                                    alt={1}
                                    animationFinished={[Function]}
                                    blockClick={false}
                                    center={
                                                      Object {
                                                                        "centerX": 0,
                                                                        "centerY": 0,
                                                                      }
                                    }
                                    code={1}
                                    playRounds={undefined}
                                    playerId={0}
                                    removeCard={[Function]}
                                    src={undefined}
                                    value={1}
                  />
</div>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <p>
                      Score:
                      0
</p>,
                    Array [
                      <Connect(Card)
                        alt={3}
                        animationFinished={[Function]}
                        blockClick={true}
                        center={
                                                Object {
                                                                        "centerX": 0,
                                                                        "centerY": 0,
                                                                      }
                        }
                        code={3}
                        playRounds={undefined}
                        playerId={2}
                        removeCard={[Function]}
                        src={undefined}
                        value={3}
/>,
                    ],
                  ],
                  "className": "top-row",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        "Score:",
                        0,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      "Score:",
                      0,
                    ],
                    "type": "p",
                  },
                  Object {
                    "instance": null,
                    "key": "3",
                    "nodeType": "class",
                    "props": Object {
                      "alt": 3,
                      "animationFinished": [Function],
                      "blockClick": true,
                      "center": Object {
                        "centerX": 0,
                        "centerY": 0,
                      },
                      "code": 3,
                      "playRounds": undefined,
                      "playerId": 2,
                      "removeCard": [Function],
                      "src": undefined,
                      "value": 3,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "player table",
                },
                "ref": Object {
                  "current": null,
                },
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <p>
                      Score:
                      0
</p>,
                    Array [
                      <Connect(Card)
                        alt={1}
                        animationFinished={[Function]}
                        blockClick={false}
                        center={
                                                Object {
                                                                        "centerX": 0,
                                                                        "centerY": 0,
                                                                      }
                        }
                        code={1}
                        playRounds={undefined}
                        playerId={0}
                        removeCard={[Function]}
                        src={undefined}
                        value={1}
/>,
                    ],
                  ],
                  "className": "bottom-row",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        "Score:",
                        0,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      "Score:",
                      0,
                    ],
                    "type": "p",
                  },
                  Object {
                    "instance": null,
                    "key": "1",
                    "nodeType": "class",
                    "props": Object {
                      "alt": 1,
                      "animationFinished": [Function],
                      "blockClick": false,
                      "center": Object {
                        "centerX": 0,
                        "centerY": 0,
                      },
                      "code": 1,
                      "playRounds": undefined,
                      "playerId": 0,
                      "removeCard": [Function],
                      "src": undefined,
                      "value": 1,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              null,
            ],
            "className": "player player-right",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            null,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
